sw/device/silicon_creator/lib/rescue/rescue_spi.c

```diff
@@ -0,0 +1,129 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/boot_data.h"
+#include "sw/device/silicon_creator/lib/dbg_print.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/spi_device.h"
+#include "sw/device/silicon_creator/lib/drivers/usb.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/rescue/dfu.h"
+#include "sw/device/silicon_creator/lib/rescue/rescue.h"
+#include "sw/device/silicon_creator/lib/rescue/sfdp.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+enum {
+  /**
+   * Base address of the spi_device registers.
+   */
+  kBase = TOP_EARLGREY_SPI_DEVICE_BASE_ADDR,
+  /**
+   * The flash buffer is a 2K region within the egress buffer, starting at the
+   * beginning of the egress buffer.
+   */
+  kFlashBuffer = 0,
+  /**
+   * The mailbox buffer is a 1K region within the egress buffer starting at 2K
+   */
+  kMailboxBuffer = 2048,
+  /**
+   * The mailbox address is located at SPI address 0x800.
+   */
+  kMailboxAddress = 0x800,
+};
+
+extern uint32_t spi_device_control(void);
+
+void dfu_transport_data(dfu_ctx_t *ctx, usb_dir_t dir, void *data, size_t len,
+                        usb_transfer_flags_t flags) {
+  if (dir & kUsbDirIn) {
+    // IN transactions get copied into the SPI egress buffer.
+    spi_device_copy_to_egress(kFlashBuffer, data, len);
+  } else {
+    // We set the total expected length for OUT transactions here.
+    // The processing of the data packets happens in the main loop in response
+    // to PAGE_PROGRAM opcodes.
+    ctx->expected_len = len;
+  }
+}
+
+rom_error_t dfu_transport_setupdata(dfu_ctx_t *ctx, usb_setup_data_t *setup) {
+  OT_DISCARD(ctx);
+  OT_DISCARD(setup);
+  // We don't handle the standard setupdata requests over SPI.
```

Comment:
Except for SetInterface and GetInterface, funnily enough. Guess you get a scratch register for fun. ;)

Reply:
I suppose I do, but only when the recipient is `Interface`.  Updated the comment.